File: leetcode/463_island_perimeter.py
from typing import List


class Solution:

    WATER, LAND = range(2)

    def islandPerimeter(self, grid: List[List[int]]) -> int:
        perimeter = 0
        if not any(grid):
            return perimeter
        rows = len(grid)
        cols = len(grid[0])

        def get_cell_perimeter(y, x):
            perimeter = 0
            for dy, dx in ((0, 1), (0, -1), (1, 0), (-1, 0)):
                yy, xx = dy+y, dx+x
                if 0 <= yy < rows and 0 <= xx < cols:
                    if grid[yy][xx] == self.WATER:
                        perimeter += 1
                else:
                    perimeter += 1
            return perimeter

        for y, row in enumerate(grid):
            for x, val in enumerate(row):
                if val == self.LAND:
                    perimeter += get_cell_perimeter(y, x)
        return perimeter

File: leetcode/java/463IslandPerimeter.java
class Solution {

    private final int LAND = 1;
    private final int EDGES_PER_CELL = 4;
    private final int EDGES_PER_NEIGHBOR_CONNECTION = 2;

    public int islandPerimeter(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int result = 0;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == LAND) {
                    result += EDGES_PER_CELL;
                    
                    if (r > 0 && grid[r-1][c] == LAND) {
                        result -= EDGES_PER_NEIGHBOR_CONNECTION;
                    }
                    
                    if (c > 0 && grid[r][c-1] == LAND) {
                        result -= EDGES_PER_NEIGHBOR_CONNECTION;
                    }
                }
            }
        }
        
        return result;
    }
}
